### ISSUE TITLE
training in different steps

### ISSUE THREAD

**Issue by alevillard** [vlfeat/matconvnet#491]
Hi, I try to finetune vggD, but I saw that if I make for example a training of 10 epochs in one time it works at the test. But if I train for example the first 5 epoch, and then stop it (for problems of time for example), and then start again to train other 5 epoch, resuming the previous five, I will see that the test fail because all the scores output are close to zero. Do you know if there is some option to add to the code for which I can stop a train and then restart from that point avoiding this problem? Maybe I see that stopping and then resuming, the code re-reinizialized the parameters and maybe that's why of this problem I have. Someone could help me? Thank you, I attached the code I use for finetuning:

 function [net, info] = vggD_train(imdb, expDir, varargin)

	vl_setupnn;

	% some common options
	opts.train.batchSize = 20;
	opts.train.numEpochs = 10 ;
	opts.train.continue = true ;
	opts.train.gpus = [1] ;
	opts.train.learningRate = 1e-3;
	opts.train.weightDecay = 2e-4;
	opts.train.momentum = 0.9;
	opts.train.expDir = expDir;
	opts.train.numSubBatches = 1;
	% getBatch options
	bopts.useGpu = numel(opts.train.gpus) >  0 ;
	opts.optMethod = 'gradient'; % ['adagrad', 'gradient']

     if(numel(varargin) > 0)
		netPre = varargin{1};
    end
% network definition!
	% MATLAB handle, passed by reference
	net = dagnn.DagNN() ;

    net.addLayer('conv1_1', dagnn.Conv('size', [3 3 3 64], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'input'}, {'conv1'},  {'conv1_1f'  'conv1_1b'});
	net.addLayer('relu1_1', dagnn.ReLU(), {'conv1'}, {'relu1'}, {});
	net.addLayer('conv1_2', dagnn.Conv('size', [3 3 64 64], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu1'}, {'conv2'},  {'conv1_2f'  'conv1_2b'});
	net.addLayer('relu1_2', dagnn.ReLU(), {'conv2'}, {'relu2'}, {});
	%net.addLayer('lrn1', dagnn.LRN('param', [5 1 0.0001/5 0.75]), {'relu1'}, {'lrn1'}, {});
	net.addLayer('pool1', dagnn.Pooling('method', 'max', 'poolSize', [2, 2], 'stride', [2 2], 'pad', [0 0 0 0]), {'relu2'}, {'pool1'}, {});
    
    net.addLayer('conv2_1', dagnn.Conv('size', [3 3 64 128], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'pool1'}, {'conv3'},  {'conv2_1f'  'conv2_1b'});
	net.addLayer('relu2_1', dagnn.ReLU(), {'conv3'}, {'relu3'}, {});
	net.addLayer('conv2_2', dagnn.Conv('size', [3 3 128 128], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu3'}, {'conv4'},  {'conv2_2f'  'conv2_2b'});
	net.addLayer('relu2_2', dagnn.ReLU(), {'conv4'}, {'relu4'}, {});
	%net.addLayer('lrn2', dagnn.LRN('param', [5 1 0.0001/5 0.75]), {'relu2'}, {'lrn2'}, {});
	net.addLayer('pool2', dagnn.Pooling('method', 'max', 'poolSize', [2, 2], 'stride', [2 2], 'pad', [0 0 0 0]), {'relu4'}, {'pool2'}, {});

    net.addLayer('conv3_1', dagnn.Conv('size', [3 3 128 256], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'pool2'}, {'conv5'},  {'conv3_1f'  'conv3_1b'});
	net.addLayer('relu3_1', dagnn.ReLU(), {'conv5'}, {'relu5'}, {});
    net.addLayer('conv3_2', dagnn.Conv('size', [3 3 256 256], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu5'}, {'conv6'},  {'conv3_2f'  'conv3_2b'});
	net.addLayer('relu3_2', dagnn.ReLU(), {'conv6'}, {'relu6'}, {});
	net.addLayer('conv3_3', dagnn.Conv('size', [3 3 256 256], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu6'}, {'conv7'},  {'conv3_3f'  'conv3_3b'});
	net.addLayer('relu3_3', dagnn.ReLU(), {'conv7'}, {'relu7'}, {});
	%net.addLayer('lrn3', dagnn.LRN('param', [5 1 0.0001/5 0.75]), {'relu2'}, {'lrn2'}, {});
	net.addLayer('pool3', dagnn.Pooling('method', 'max', 'poolSize', [2, 2], 'stride', [2 2], 'pad', [0 0 0 0]), {'relu7'}, {'pool3'}, {});
    
    net.addLayer('conv4_1', dagnn.Conv('size', [3 3 256 512], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'pool3'}, {'conv8'},  {'conv4_1f'  'conv4_1b'});
	net.addLayer('relu4_1', dagnn.ReLU(), {'conv8'}, {'relu8'}, {});
	net.addLayer('conv4_2', dagnn.Conv('size', [3 3 512 512], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu8'}, {'conv9'},  {'conv4_2f'  'conv4_2b'});
	net.addLayer('relu4_2', dagnn.ReLU(), {'conv9'}, {'relu9'}, {});
	net.addLayer('conv4_3', dagnn.Conv('size', [3 3 512 512], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu9'}, {'conv10'},  {'conv4_3f'  'conv4_3b'});
	net.addLayer('relu4_3', dagnn.ReLU(), {'conv10'}, {'relu10'}, {});
	%net.addLayer('lrn4', dagnn.LRN('param', [5 1 0.0001/5 0.75]), {'relu2'}, {'lrn2'}, {});
	net.addLayer('pool4', dagnn.Pooling('method', 'max', 'poolSize', [2, 2], 'stride', [2 2], 'pad', [0 0 0 0]), {'relu10'}, {'pool4'}, {});
  
    net.addLayer('conv5_1', dagnn.Conv('size', [3 3 512 512], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'pool4'}, {'conv11'},  {'conv5_1f'  'conv5_1b'});
	net.addLayer('relu5_1', dagnn.ReLU(), {'conv11'}, {'relu11'}, {});
	net.addLayer('conv5_2', dagnn.Conv('size', [3 3 512 512], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu11'}, {'conv12'},  {'conv5_2f'  'conv5_2b'});
	net.addLayer('relu5_2', dagnn.ReLU(), {'conv12'}, {'relu12'}, {});
	net.addLayer('conv5_3', dagnn.Conv('size', [3 3 512 512], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu12'}, {'conv13'},  {'conv5_3f'  'conv5_3b'});
	net.addLayer('relu5_3', dagnn.ReLU(), {'conv13'}, {'relu13'}, {});
	net.addLayer('pool5', dagnn.Pooling('method', 'max', 'poolSize', [2, 2], 'stride', [2 2], 'pad', [0 0 0 0]), {'relu13'}, {'pool5'}, {});

	net.addLayer('conv6', dagnn.Conv('size', [1 1 512 4096], 'hasBias', true, 'stride', [1, 1], 'pad', [0 0 0 0]), {'pool5'}, {'conv14'},  {'conv6f'  'conv6b'});
	net.addLayer('relu6', dagnn.ReLU(), {'conv14'}, {'relu14'}, {});
	net.addLayer('drop6', dagnn.DropOut('rate', 0.5), {'relu14'}, {'drop6'}, {});

	net.addLayer('conv7', dagnn.Conv('size', [1 1 4096 4096], 'hasBias', true, 'stride', [1, 1], 'pad', [0 0 0 0]), {'drop6'}, {'conv15'},  {'conv7f'  'conv7b'});
	net.addLayer('relu7', dagnn.ReLU(), {'conv15'}, {'relu15'}, {});
	net.addLayer('drop7', dagnn.DropOut('rate', 0.5), {'relu15'}, {'drop7'}, {});

	net.addLayer('classifier', dagnn.Conv('size', [1 1 4096 263], 'hasBias', true, 'stride', [1, 1], 'pad', [0 0 0 0]), {'drop7'}, {'classifier'},  {'conv8f'  'conv8b'});
	net.addLayer('prob', dagnn.SoftMax(), {'classifier'}, {'prob'}, {});
	net.addLayer('objective', dagnn.Loss('loss', 'log'), {'prob', 'label'}, {'objective'}, {});
	net.addLayer('error', dagnn.Loss('loss', 'classerror'), {'prob','label'}, 'error') ;
	% -- end of the network

    % initialization of the weights (CRITICAL!!!!)
    if(numel(varargin) > 0)
		initNet_FineTuning(net, netPre);
        
	else
		 initNet_He(net);
	end	
   
    %train
    info = cnn_train_dag(net, imdb, @(i,b) getBatchDisk(bopts,i,b), opts.train, 'val', find(imdb.images.set == 2)) ;
end

function initNet_FineTuning(net, netPre)
	net.initParams();

	ind = 1;
	for l=1:length(net.layers)-4
		% is a convolution layer?
		if(strcmp(class(net.layers(l).block), 'dagnn.Conv'))
			f_ind = net.layers(l).paramIndexes(1);
			b_ind = net.layers(l).paramIndexes(2);
            
            net.params(f_ind).value = netPre.params(ind).value
			ind = ind + 1;
			net.params(b_ind).value = netPre.params(ind).value
			ind = ind + 1;
          if l<= length(net.layers)-10  %freeze the pararamiters
			net.params(f_ind).learningRate = 0;
			net.params(f_ind).weightDecay = 1;
			net.params(b_ind).learningRate = 0;
			net.params(b_ind).weightDecay = 1;
			
          else  %let paramiters to change in last layers
            net.params(f_ind).learningRate = 1;
			net.params(f_ind).weightDecay = 1;
			net.params(b_ind).learningRate = 1;
			net.params(b_ind).weightDecay = 1;   
          end
		end
	end

	for l=length(net.layers)-3:length(net.layers)
		% is a convolution layer?
		if(strcmp(class(net.layers(l).block), 'dagnn.Conv'))
			f_ind = net.layers(l).paramIndexes(1)
			b_ind = net.layers(l).paramIndexes(2)

            
			[h,w,in,out] = size(net.params(f_ind).value)
			he_gain = 1e-2*sqrt(2/(h*w*in)) % sqrt(1/fan_in)
			net.params(f_ind).value = (1/5)*he_gain*randn(size(net.params(f_ind).value), 'single');
			net.params(f_ind).learningRate = 1;
			net.params(f_ind).weightDecay = 1;

			net.params(b_ind).value = zeros(size(net.params(b_ind).value), 'single');
			net.params(b_ind).learningRate = 0.5;
			net.params(b_ind).weightDecay = 1;
			
		end
	end
end

% getBatch for IMDBs that are too big to be in RAM
function inputs = getBatchDisk(opts, imdb, batch)
	cell_images = vl_imreadjpeg(imdb.images.filenames(batch),'numThreads', 2, 'Resize', [224 224]) ;
	%[H, W, CH] = size(cell_images{1});

	images = zeros(224, 224, 3, 'single');
    %size(images)
	for i=1:numel(cell_images)
		im = cell_images{i};
        %im=imresize(im, [224 224]);
        %size(im)
		images(:,:,:, i) = im - imdb.images.data_mean;
	end

	labels = imdb.images.labels(1,batch) ;
	if opts.useGpu > 0
  		images = gpuArray(images) ;
	end

	inputs = {'input', images, 'label', labels} ;
end

function initNet_He(net)
	net.initParams();
	for l=1:length(net.layers)
		% is a convolution layer?
		if(strcmp(class(net.layers(l).block), 'dagnn.Conv'))
			f_ind = net.layers(l).paramIndexes(1);
			b_ind = net.layers(l).paramIndexes(2);

			[h,w,in,out] = size(net.params(f_ind).value);
			he_gain = 1e-2*sqrt(2/(h*w*in)); % sqrt(1/fan_in)
			net.params(f_ind).value = (1/2)*he_gain*randn(size(net.params(f_ind).value), 'single')
			net.params(f_ind).learningRate = 1;
			net.params(f_ind).weightDecay = 1;

			net.params(b_ind).value = zeros(size(net.params(b_ind).value), 'single');
			net.params(b_ind).learningRate = 0.5;
			net.params(b_ind).weightDecay = 1;
		end
	end
end

**Comment by albanie** [comment]
The code looks OK. By setting `opts.train.continue = true`, training should resume using the weights that were found just before you stopped.  

The code responsible for resuming training correctly starts around [line 73](https://github.com/vlfeat/matconvnet/blob/master/examples/cnn_train_dag.m#L73-L77) in `cnn_train_dag`:

    start = opts.continue * findLastCheckpoint(opts.expDir) ;
    if start >= 1
      fprintf('%s: resuming by loading epoch %d\n', mfilename, start) ;
      [net, stats] = loadState(modelPath(start)) ;
    end

Could you try adding a breakpoint to check the value of `start` when you resume training after 5 epochs? It should be 5, but if it's zero then training will start from scratch.

**Comment by lenck** [comment]
Hmm, yeah, this is strange. There really should not be any difference, as before any epochs the momentum is being reset as well, so the solver state is not important... 

Maybe when you start training again you do not set the correct learning rate? Note that using the `logspace` is a bit tricky for this...

**Comment by alevillard** [comment]
I don't know, I tried with a database of 50 classes, training in two steps and I have the same problem: the output of 'prob' is always around 1/NumberofClasses, so in this case 0.02. if I use 263 classes I have all scores at 0.0038.

This are for example the scores I have in a prediction. 
0.019998210
0.019981256
0.020128585
0.020023547
0.020005781
0.020002332
0.020133033
0.020359971
0.019874524
0.020124212
0.019910786
0.020272855
0.020152502
0.020056034
0.019909369
0.019722419
0.020032471
0.020231700
0.019959165
0.020066734
0.019950001
0.019663915
0.020058043
0.019582516
0.019890025
0.020117616
0.019984225
0.019828182
0.019935135
0.020031612
0.019910730
0.019999595
0.020000188
0.019994894
0.020070806
0.020190164
0.020037146
0.019939467
0.019911716
0.020211998
0.019910039
0.019992314
0.020069361
0.020094782
0.020098317
0.019718209
0.019900616
0.019852782
0.019980807
0.020129286

**Comment by alevillard** [comment]
I have used also another architecture, Googlenet, but the results are the same. @lenck My learning rate is always fixed at 1e-3.

**Comment by lenck** [comment]
Hmm, just a silly question but when you start the second step fine-tuning, don't you reinitialize the top layer to random values? Check that the saved network is the same one which being loaded and used in the second step...
